pkg/shp/cmd/follower/follow.go

```diff
@@ -0,0 +1,229 @@
+
+	enteredRunningState bool
+}
+
+// NewFollower returns a Follower instance
+func NewFollower(
+	ctx context.Context,
+	buildRunName string,
+	ioStreams *genericclioptions.IOStreams,
+	params *params.Params,
+) (*Follower, error) {
+	follower := &Follower{ctx: ctx, ioStreams: ioStreams, buildRunName: buildRunName, logLock: sync.Mutex{}, tailLogsStarted: make(map[string]bool)}
+	follower.complete(params)
+	return follower, nil
+}
+
+// GetLogLock returns the mutex used for coordinating access to log buffers.
+func (f *Follower) GetLogLock() *sync.Mutex {
+	return &f.logLock
+}
+
+func (f *Follower) complete(params *params.Params) error {
+	clientset, err := params.ClientSet()
+	if err != nil {
+		return err
+	}
+	f.logTail = tail.NewTail(f.ctx, clientset)
+	f.namespace = params.Namespace()
+	if f.shpClientset, err = params.ShipwrightClientSet(); err != nil {
+		return err
+	}
+	f.kclientset, err = params.ClientSet()
+	if err != nil {
+		return err
+	}
+	to, err := params.RequestTimeout()
+	if err != nil {
+		return err
+	}
+	f.pw, err = reactor.NewPodWatcher(f.ctx, to, f.kclientset, params.Namespace())
+	if err != nil {
+		return err
+	}
+
+	f.pw.WithOnPodModifiedFn(f.OnEvent)
+	f.pw.WithTimeoutPodFn(f.OnTimeout)
+	f.pw.WithNoPodEventsYetFn(f.OnNoPodEventsYet)
+	return nil
+}
+
+// Log prints a message
+func (f *Follower) Log(msg string) {
+	// concurrent fmt.Fprintf(r.ioStream.Out...) calls need locking to avoid data races, as we 'write' to the stream
+	f.logLock.Lock()
+	defer f.logLock.Unlock()
+	fmt.Fprint(f.ioStreams.Out, msg)
+}
+
+// tailLogs start tailing logs for each container name in init-containers and containers, if not
+// started already.
+func (f *Follower) tailLogs(pod *corev1.Pod) {
+	containers := append(pod.Spec.InitContainers, pod.Spec.Containers...)
+	for _, container := range containers {
+		if _, exists := f.tailLogsStarted[container.Name]; exists {
+			continue
+		}
+		f.tailLogsStarted[container.Name] = true
+		f.logTail.Start(pod.GetNamespace(), pod.GetName(), container.Name)
+	}
+}
+
+// stop invoke stop on streaming components.
+func (f *Follower) stop() {
+	f.logTail.Stop()
+	f.pw.Stop()
+}
+
+// OnEvent reacts on pod state changes, to start and stop tailing container logs.
+func (f *Follower) OnEvent(pod *corev1.Pod) error {
+	switch pod.Status.Phase {
+	case corev1.PodRunning:
+		if !f.enteredRunningState {
+			f.Log(fmt.Sprintf("Pod %q in %q state, starting up log tail", pod.GetName(), corev1.PodRunning))
+			f.enteredRunningState = true
+			// graceful time to wait for container start
+			time.Sleep(3 * time.Second)
+			// start tailing container logs
+			f.tailLogs(pod)
+		}
+	case corev1.PodFailed:
+		msg := ""
+		br, err := f.shpClientset.ShipwrightV1alpha1().BuildRuns(pod.Namespace).Get(f.ctx, f.buildRunName, metav1.GetOptions{})
+		switch {
+		case err == nil && br.IsCanceled():
+			msg = fmt.Sprintf("BuildRun %q has been canceled.\n", br.Name)
+		case err == nil && br.DeletionTimestamp != nil:
```

Comment:
Is this needed: ?

```suggestion
		case (err == nil && br.DeletionTimestamp != nil) || (err != nil && k8serrors.IsNotFound(err)):
```

Reply:
sure falls under another race condition with the build controller if a build failed and then somebody detects that and issues deletes

I'll add